src/drivers/fluid_winmidi.c

```diff
@@ -224,115 +370,178 @@ new_fluid_winmidi_driver(fluid_settings_t *settings,
         return NULL;
     }
 
-    /* get the device name. if none is specified, use the default device. */
-    if(fluid_settings_copystr(settings, "midi.winmidi.device", dev_name, MAXPNAMELEN) != FLUID_OK)
+    /* check if there is any midi devices installed */
+    /* get the maximum number of devices to handle */
+    fluid_settings_getint(settings, "midi.winmidi.maxdevices", &max_devices);
+    if(max_devices < 0)
     {
-        FLUID_LOG(FLUID_DBG, "No MIDI in device selected, using \"default\"");
-        FLUID_STRCPY(dev_name, "default");
+        max_devices = 1;
     }
+    num = midiInGetNumDevs(); /* get number of real devices installed */
 
-    /* check if there any midi devices installed */
-    num = midiInGetNumDevs();
+    if(num < max_devices)
+    {
+        FLUID_LOG(FLUID_ERR, "not enough MIDI in devices found. Expected:%d found:%d",
+                  max_devices, num);
+        return NULL;
+    }
+
+    /* allocation of driver structure size dependant of max_devices */
+    i = sizeof(fluid_winmidi_driver_t) + (max_devices - 1) * sizeof(device_infos);
+    dev = FLUID_MALLOC(i);
 
-    if(num == 0)
+    if(dev == NULL)
     {
-        FLUID_LOG(FLUID_ERR, "no MIDI in devices found");
         return NULL;
     }
 
-    /* find the device */
-    if(FLUID_STRCASECMP("default", dev_name) != 0)
+    FLUID_MEMSET(dev, 0, i); /* reset structure members */
+
+    /* parse device name */
+    /* get the device name. if none is specified, use the default device. */
+    if(fluid_settings_copystr(settings, "midi.winmidi.device", dev_name, MAXPNAMELEN) != FLUID_OK)
     {
-        for(i = 0; i < num; i++)
+        FLUID_LOG(FLUID_DBG, "No MIDI in device selected, using \"default\"");
+        FLUID_STRCPY(dev_name, "default");
+    }
+
+    /* look if the device name starts with the prefix 'multi:'. */
+    if( FLUID_STRNCASECMP(multi_dev_name, dev_name, MULTI_DEV_PREFIX_LEN) == 0)
+    {
+        /* multi devices name "multi:x[ y z ..]". parse devices index: x,y,..
+          Each ascii index are separated by space(s) caracters.
+        */
+        int dev_idx;               /* device index */
+        char *cur_idx, *next_idx;  /* current and next ascii index pointer */
+        next_idx = &dev_name[MULTI_DEV_PREFIX_LEN];
+        do
         {
-            res = midiInGetDevCaps(i, &in_caps, sizeof(MIDIINCAPS));
+            /* try to convert next ascii index */
+            cur_idx = next_idx;
+            dev_idx = g_ascii_strtoll(cur_idx, &next_idx, 10);
+            if (cur_idx == next_idx                /* not an integer number */
```

Comment:
If there are two spaces in the string the parsing would fail, wouldn't it? To make it more robust, how about using `fluid_strtok()`?

Reply:
This could be done. I will see that.